File: firebase/functions/lib/utils/ics_util.dart
@JS()
library ics;

import 'package:enum_to_string/enum_to_string.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'firestore_utils.dart';
import 'package:data_models/firestore/event.dart';
import 'package:data_models/firestore/community.dart';
import 'package:node_interop/node.dart';
import 'package:js/js.dart';

final _icsProdId = functions.config.get('ics.prod_id') as String;

IcsLib requireIcsLib() {
  return require('ics') as IcsLib;
}

@JS()
@anonymous
abstract class IcsLib {
  external IcsResult createEvents(List<IcsEvent> events);
}

@JS()
@anonymous
abstract class IcsResult {
  external ErrorResult? get error;
  external String? get value;
}

@JS()
@anonymous
abstract class ErrorResult {
  external List<dynamic>? get errors;
}

@JS()
@anonymous
class IcsEvent {
  external List<int> get start;
  external String get startInputType;
  external IcsDuration get duration;
  external String title;
  external String url;
  external String uid;
  external String productId;

  external factory IcsEvent({
    List<dynamic> start,
    String startInputType,
    IcsDuration duration,
    String title,
    String url,
    String uid,
    String productId,
  });
}

@JS()
@anonymous
class IcsDuration {
  external int get minutes;

  external factory IcsDuration({
    int minutes,
  });
}

final emptyIcs = '''BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
PRODID:$_icsProdId
METHOD:PUBLISH
X-PUBLISHED-TTL:PT1H
END:VCALENDAR
''';

final icsUtil = IcsUtil();

class IcsUtil {
  /// Return an ICS string containing future events for a given community
  Future<String> getIcsForUpcomingEvents({required Community community}) async {
    List<Event> events = (await firestore
            .collectionGroup('events')
            .where(Event.kFieldCommunityId, isEqualTo: community.id)
            .where(Event.kFieldIsPublic, isEqualTo: true)
            .where(
              Event.kFieldStatus,
              isEqualTo: EnumToString.convertToString(EventStatus.active),
            )
            .where(
              Event.kFieldScheduledTime,
              isGreaterThanOrEqualTo: DateTime.now(),
            )
            .orderBy(Event.kFieldScheduledTime)
            .get())
        .documents
        .map(
          (doc) => Event.fromJson(
            firestoreUtils.fromFirestoreJson(doc.data.toMap()),
          ),
        )
        .toList();

    final domain = functions.config.get('app.domain') as String;

    final scheduledEvents =
        events.where((event) => event.scheduledTime != null).map((event) {
      final scheduledTime = event.scheduledTime!.toUtc();
      return IcsEvent(
        title: event.title ?? 'Event',
        url:
            'https://$domain/space/${community.displayId}/discuss/${event.templateId}/${event.id}',
        uid: event.id,
        startInputType: 'utc',
        start: [
          scheduledTime.year,
          scheduledTime.month,
          scheduledTime.day,
          scheduledTime.hour,
          scheduledTime.minute,
        ],
        duration: IcsDuration(minutes: 60),
        productId: _icsProdId,
      );
    }).toList();

    if (scheduledEvents.isEmpty) {
      return emptyIcs;
    } else {
      final ics = requireIcsLib();
      final result = ics.createEvents(scheduledEvents);

      final error = result.error;
      final value = result.value;
      if (error != null) {
        print(error.errors);
        throw Exception('Could not generate feed');
      }
      if (value == null) {
        throw Exception('Could not generate feed');
      }
      return value;
    }
  }
}

File: firebase/functions/src/utils/ics_util.ts
import * as functions from 'firebase-functions'
import { firestore, firestoreUtils } from './infra/firestore_utils'
import { Community, Event, EventStatus } from '../types'

const ics = require('ics') as {
    createEvents: (events: IcsEventInput[]) => { error: unknown; value?: string }
}

interface IcsEventInput {
    start: number[]
    startInputType: string
    duration: { minutes: number }
    title: string
    url: string
    uid: string
    productId: string
}

const icsProdId = () => (functions.config().ics?.prod_id as string) ?? '-//frankly//EN'

const emptyIcs = () => `BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
PRODID:${icsProdId()}
METHOD:PUBLISH
X-PUBLISHED-TTL:PT1H
END:VCALENDAR
`

export const icsUtil = {
    async getIcsForUpcomingEvents({ community }: { community: Community }): Promise<string> {
        const now = new Date()
        const domain = functions.config().app?.domain as string

        const eventDocs = await firestore
            .collectionGroup('events')
            .where('communityId', '==', community.id)
            .where('isPublic', '==', true)
            .where('status', '==', EventStatus.active)
            .where('scheduledTime', '>=', now)
            .orderBy('scheduledTime')
            .get()

        const events = eventDocs.docs.map((doc) => ({
            ...(firestoreUtils.fromFirestoreJson(doc.data()) as unknown as Event),
            id: doc.id,
        }))

        const scheduledEvents = events
            .filter((e) => e.scheduledTime != null)
            .map((event) => {
                const t = event.scheduledTime!
                const d = t instanceof Date ? t : new Date(t)
                return {
                    title: event.title ?? 'Event',
                    url: `https://${domain}/space/${
                        community.displayIds?.[0] ?? community.id
                    }/discuss/${event.templateId}/${event.id}`,
                    uid: event.id,
                    startInputType: 'utc',
                    start: [
                        d.getUTCFullYear(),
                        d.getUTCMonth() + 1,
                        d.getUTCDate(),
                        d.getUTCHours(),
                        d.getUTCMinutes(),
                    ],
                    duration: { minutes: 60 },
                    productId: icsProdId(),
                } as IcsEventInput
            })

        if (scheduledEvents.length === 0) {
            return emptyIcs()
        }

        const result = ics.createEvents(scheduledEvents)
        if (result.error) {
            console.error('ICS error:', result.error)
            throw new Error('Could not generate feed')
        }
        if (!result.value) {
            throw new Error('Could not generate feed')
        }
        return result.value
    },
}
